### PR TITLE
Added multiqc-tools typescript expressions and added samplename change to cl-config in multiqc for wts workflow

### PR DESCRIPTION
**Pull request by alexiswl**
Added function add_sample_to_sample_name_replace_in_multiqc_cl_config to multiqc-tools ts expressions and added tests for said function

Fixes https://github.com/umccr/cwl-ica/issues/280

### CODE REVIEW

**Comment by alexiswl** [comment]
#  wfr.4c02d99cf96f4d1b8e4af33274e5933d 

## Inputs


<details>
<summary>Details</summary>

```
{
  "annotation_file": {
    "class": "File",
    "location": "gds://development/reference-data/dragen_wts/hg38/ref-transcripts.non-zero-length.gtf"
  },
  "blacklist": {
    "class": "File",
    "location": "gds://development/reference-data/dragen_wts/arriba/2.3.0/blacklist_hg38_GRCh38_v2.3.0.tsv.gz"
  },
  "cl_config": "{\"sample_names_replace\":{\"PRJ220412\":\"Ref_1_Good\",\"MDX210318\":\"Ref_2_Good\",\"PRJ220466\":\"Ref_3_Good\",\"PRJ211234\":\"Ref_4_Bad\",\"PRJ220790\":\"Ref_5_Bad\",\"L2200121_dragen_qualimap\":\"Ref_1_Good\",\"L2101521_dragen_qualimap\":\"Ref_2_Good\",\"L2200188_dragen_qualimap\":\"Ref_3_Good\",\"L2101763_dragen_qualimap\":\"Ref_4_Bad\",\"L2200311_dragen_qualimap\":\"Ref_5_Bad\"}}",
  "cytobands": {
    "class": "File",
    "location": "gds://development/reference-data/dragen_wts/arriba/2.3.0/cytobands_hg38_GRCh38_v2.3.0.tsv"
  },
  "enable_duplicate_marking": false,
  "enable_map_align_output": true,
  "enable_sort": true,
  "fastq_list_rows": [
    {
      "lane": 1,
      "read_1": {
        "class": "File",
        "location": "gds://production/primary_data/211125_A00130_0185_AHWC2HDSX2/2022010575cdcf71/WTS_NebRNA/MDX210377_L2101522_S2_L004_R1_001.fastq.gz"
      },
      "read_2": {
        "class": "File",
        "location": "gds://production/primary_data/211125_A00130_0185_AHWC2HDSX2/2022010575cdcf71/WTS_NebRNA/MDX210377_L2101522_S2_L004_R2_001.fastq.gz"
      },
      "rgid": "MDX210377",
      "rglb": "L2101522",
      "rgsm": "MDX210377"
    }
  ],
  "output_directory": "L2101522_dragen",
  "output_file_prefix": "MDX210377",
  "protein_domains": {
    "class": "File",
    "location": "gds://development/reference-data/dragen_wts/arriba/hg38/arriba-protein-domains.gff3"
  },
  "qc_reference_samples": [
    {
      "class": "Directory",
      "location": "gds://development/reference-data/dragen_wts/multiqc/2022-09-14--3.9.3/SBJ01563"
    },
    {
      "class": "Directory",
      "location": "gds://development/reference-data/dragen_wts/multiqc/2022-09-14--3.9.3/SBJ01147"
    },
    {
      "class": "Directory",
      "location": "gds://development/reference-data/dragen_wts/multiqc/2022-09-14--3.9.3/SBJ01620"
    },
    {
      "class": "Directory",
      "location": "gds://development/reference-data/dragen_wts/multiqc/2022-09-14--3.9.3/SBJ01286"
    },
    {
      "class": "Directory",
      "location": "gds://development/reference-data/dragen_wts/multiqc/2022-09-14--3.9.3/SBJ01673"
    }
  ],
  "reference_fasta": {
    "class": "File",
    "location": "gds://development/reference-data/genomes/hg38/hg38.fa"
  },
  "reference_tar": {
    "class": "File",
    "location": "gds://development/reference-data/dragen_hash_tables/v8/hg38/altaware-cnv-anchored/hg38-v8-altaware-cnv-anchored.tar.gz"
  }
}
```

</details>


## Multiqc Tasks

```
ica tasks runs get trn.7a86e5d16089413ca2777a6fb34d9c08 -o json
```

### Workflow Input CL Config

```
ica workflows runs get wfr.4c02d99cf96f4d1b8e4af33274e5933d -o json | jq '.input | fromjson | .cl_config | fromjson'
```

```json
{
  "sample_names_replace": {
    "PRJ220412": "Ref_1_Good",
    "MDX210318": "Ref_2_Good",
    "PRJ220466": "Ref_3_Good",
    "PRJ211234": "Ref_4_Bad",
    "PRJ220790": "Ref_5_Bad",
    "L2200121_dragen_qualimap": "Ref_1_Good",
    "L2101521_dragen_qualimap": "Ref_2_Good",
    "L2200188_dragen_qualimap": "Ref_3_Good",
    "L2101763_dragen_qualimap": "Ref_4_Bad",
    "L2200311_dragen_qualimap": "Ref_5_Bad"
  }
}
```

### Task CL-Config

```
ica tasks runs get trn.7a86e5d16089413ca2777a6fb34d9c08 -o json | jq '.execution | .args[2] | fromjson'
```

```json
{
  "sample_names_replace": {
    "PRJ220412": "Ref_1_Good",
    "MDX210318": "Ref_2_Good",
    "PRJ220466": "Ref_3_Good",
    "PRJ211234": "Ref_4_Bad",
    "PRJ220790": "Ref_5_Bad",
    "L2200121_dragen_qualimap": "Ref_1_Good",
    "L2101521_dragen_qualimap": "Ref_2_Good",
    "L2200188_dragen_qualimap": "Ref_3_Good",
    "L2101763_dragen_qualimap": "Ref_4_Bad",
    "L2200311_dragen_qualimap": "Ref_5_Bad",
    "MDX210377_qualimap": "MDX210377"
  }
}
```

File: typescript-expressions/multiqc-tools/1.0.0/tests/multiqc-tools__1.0.0.test.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    "another_cl_config_parameter": "foo"
+});
+var MULTIQC_CL_CONFIG_STR_WITHOUT_SAMPLENAME_REPLACE = JSON.stringify({
+    "another_cl_config_parameter": "foo"
+});
+var SAMPLE_NAME_OLD = "L13243546";
+var SAMPLE_NAME_NEW = "Ref_2_Bad";
+var EXPECTED_OUTPUT_MULTIQC_CL_CONFIG_STR_WITH_SAMPLENAME_REPLACE_OBJ = {
+    "sample_names_replace": {
+        "L1234567": "Ref_1_Good"
```
Comment (skanwal) [inline typescript-expressions/multiqc-tools/1.0.0/tests/multiqc-tools__1.0.0.test.js:21]: Should this be "Ref_2_Bad" as in line18?

File: workflows/dragen-transcriptome-pipeline/3.9.3/dragen-transcriptome-pipeline__3.9.3.cwl [hunk 2]
```diff
@@ -321,7 +323,13 @@ steps:
...
         source: create_dummy_file_step/dummy_file_output
       cl_config:
-        source: cl_config
+        source:
+          - cl_config
+          - output_file_prefix
+        valueFrom: |
+          ${
+            return add_sample_to_sample_name_replace_in_multiqc_cl_config(self[0], self[1] + "_qualimap", self[1]);
+          }
```
Comment (skanwal) [inline workflows/dragen-transcriptome-pipeline/3.9.3/dragen-transcriptome-pipeline__3.9.3.cwl:332]: was this tested on ICA?

Comment (alexiswl) [reply workflows/dragen-transcriptome-pipeline/3.9.3/dragen-transcriptome-pipeline__3.9.3.cwl:332]: Yep, see this comment  - https://github.com/umccr/cwl-ica/pull/281#issuecomment-1433932756

File: typescript-expressions/multiqc-tools/1.0.0/tests/multiqc-tools__1.0.0.test.js (hunk 1 above)
Comment (alexiswl) [reply typescript-expressions/multiqc-tools/1.0.0/tests/multiqc-tools__1.0.0.test.js:21]: No, the EXPECTED_OUTPUT_MULTIQC_CL_CONFIG_STR_WITH_SAMPLENAME_REPLACE_OBJ is 

```json
{
  "sample_names_replace": {
    "L1234567": "Ref_1_Good",
    "L13243546": "Ref_2_Bad"
  }
}
```

However if we were to place these variables in directly (which is what we'd do in typescript), we'd get an issue in ES5.1 where

```json
{
  "sample_names_replace": {
    "L1234567": "Ref_1_Good",
    SAMPLE_NAME_OLD: SAMPLE_NAME_NEW 
  }
}
```

would be evaluated to 

```json
{
  "sample_names_replace": {
    "L1234567": "Ref_1_Good",
    "SAMPLE_NAME_OLD": "SAMPLE_NAME_NEW"
  }
}
```

which is why we split the initialisation of the object into https://github.com/umccr/cwl-ica/pull/281/files/a0960f3f44277857ba3de094cc90dc28c8883f1c#diff-343a61231e67f24ca51d65ca1105f653c54d59b8fd4b8dc0e1eb424d2ae73002R19 AND then append the `"L13243546": "Ref_2_Bad"` here: https://github.com/umccr/cwl-ica/pull/281/files/a0960f3f44277857ba3de094cc90dc28c8883f1c#diff-343a61231e67f24ca51d65ca1105f653c54d59b8fd4b8dc0e1eb424d2ae73002R25

Comment (skanwal) [reply typescript-expressions/multiqc-tools/1.0.0/tests/multiqc-tools__1.0.0.test.js:21]: Okay makes sense. Trying to understand what is the added value of this TS function as cl_config input to workflow is still the same and it's being passed on to this function along with output prefix.

Comment (alexiswl) [reply typescript-expressions/multiqc-tools/1.0.0/tests/multiqc-tools__1.0.0.test.js:21]: See screenshot above, as discussed on Wednesday, previously the sample name would have been "MDX210377_qualimap" but is now just "MDX210377".

**Comment by skanwal** [comment]
For record and others to follow - this will enable us to name qualimap output directory like [output_file_prefix_qualimap](https://github.com/umccr/cwl-ica/blob/main/workflows/dragen-transcriptome-pipeline/3.9.3/dragen-transcriptome-pipeline__3.9.3.cwl#L283) on GDS (maintains traceability and consistency between our workflow output dirs) but use only `output_file_prefix`  in qualimap plots in multiQC (as seen in above screenshot), opposed to one below:

Screen Shot 2023-02-21 at 12 11 54 pm

File: typescript-expressions/multiqc-tools/1.0.0/tests/multiqc-tools__1.0.0.test.js (hunk 1 above)
Comment (skanwal) [reply typescript-expressions/multiqc-tools/1.0.0/tests/multiqc-tools__1.0.0.test.js:21]: Noted and added a summary statement to this PR.

**Review by skanwal — APPROVED** [review]

**Pull request merged**